### PR TITLE
Implemented Rachford-Rice and cubic EOS

### PR DESCRIPTION
**Pull request by francoishamon**
This PR adds the first building blocks of future compositional fluid calculations: 
- Solving the Rachford-Rice equation
- Computing EOS calculation

These building blocks are implemented as free input/output functions with two unit tests to validate the results.

The other building blocks (negative two-phase flash, initial K-values, Soreide-Whitson, properties) will be added in other PRs

Ready for review

### CODE REVIEW

File: src/coreComponents/constitutive/fluid/RachfordRice.cpp [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+real64
+RachfordRice::evaluate( arrayView1d< real64 const > const kValues,
+                        arrayView1d< real64 const > const feed,
+                        arrayView1d< integer const > const presentComponentIds,
+                        real64 const & x )
+{
+  real64 value = 0.0;
+  for( integer ic = 0; ic < presentComponentIds.size(); ++ic )
```
Comment (dkachuma) [inline src/coreComponents/constitutive/fluid/RachfordRice.cpp:160]: Should be `for (integer const ic : presentComponentIds)` i.e. loop over the actual values of the indicator array.

File: src/coreComponents/constitutive/fluid/RachfordRice.cpp [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+real64
+RachfordRice::evaluateDerivative( arrayView1d< real64 const > const kValues,
+                                  arrayView1d< real64 const > const feed,
+                                  arrayView1d< integer const > const presentComponentIds,
+                                  real64 const & x )
+{
+  real64 value = 0.0;
+  for( integer ic = 0; ic < presentComponentIds.size(); ++ic )
```
Comment (dkachuma) [inline src/coreComponents/constitutive/fluid/RachfordRice.cpp:175]: Same as above

File: src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+    real64 zMin =  LvArray::NumericLimits< real64 >::max;
+    real64 zMax = -LvArray::NumericLimits< real64 >::max;
+
+    for( integer i = 0; i < 3; ++i )
+    {
+      // skip unphysical roots
+      if( roots[i] > bMixtureCoefficient )
+      {
+        // choose the root according to Gibbs' free energy minimization
+        if( zMin < roots[i] )
```
Comment (dkachuma) [inline src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp:340]: The signs are reversed here I think.

File: src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+    }
+
+    stackArray1d< real64, maxNumComps > logFugacityCoefficientsMax( numComps );
+    stackArray1d< real64, maxNumComps > logFugacityCoefficientsMin( numComps );
+    computeLogFugacityCoefficients( numComps, composition, binaryInteractionCoefficients, zMin,
+                                    aPureCoefficient, bPureCoefficient, aMixtureCoefficient, bMixtureCoefficient,
+                                    logFugacityCoefficientsMin.toSlice() );
+    computeLogFugacityCoefficients( numComps, composition, binaryInteractionCoefficients, zMax,
+                                    aPureCoefficient, bPureCoefficient, aMixtureCoefficient, bMixtureCoefficient,
+                                    logFugacityCoefficientsMin.toSlice() );
```
Comment (dkachuma) [inline src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp:358]: Second call should pass `logFugacityCoefficientsMax.toSlice()`

File: src/coreComponents/constitutive/fluid/RachfordRice.cpp (hunk 1 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/fluid/RachfordRice.cpp]: Thanks for the catch, this is fixed now

File: src/coreComponents/constitutive/fluid/RachfordRice.cpp (hunk 2 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/fluid/RachfordRice.cpp]: Thanks, also fixed

File: src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp (hunk 4 above)
Comment (francoishamon) [reply src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp]: Done

File: src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp (hunk 3 above)
Comment (dkachuma) [reply src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp]: I meant shouldn't it be `if (zMin > roots[i])` and similarly for `zMax`.

Comment (francoishamon) [reply src/coreComponents/constitutive/fluid/CubicEOSPhaseModel.hpp]: Yes!!! Thanks!

**Review by dkachuma — APPROVED** [review]

**Pull request merged**